File: app/src/main/java/mb00/android/codehub/ui/repo/adapter/CodeAdapter.java
package mb00.android.codehub.ui.repo.adapter;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.List;

import mb00.android.codehub.R;
import mb00.android.codehub.api.model.Code;
import mb00.android.codehub.logic.utils.FileSizeParser;
import mb00.android.codehub.data.BundleKeys;
import mb00.android.codehub.ui.repo.view.RepoCodeFragment;
import mb00.android.codehub.ui.repo.view.RepoFileActivity;

/**
 * A RecyclerView adapter used to display folders and files in {@link RepoCodeFragment}
 */

public class CodeAdapter extends RecyclerView.Adapter<CodeAdapter.CodeHolder> {
    
    private List<Code> fileList;
    private RecyclerView fileRecyclerView;
    private String header;
    private String user;
    private String repo;

    public CodeAdapter(List<Code> fileList, RecyclerView fileRecyclerView, String header, String user, String repo) {
        this.fileList = fileList;
        this.fileRecyclerView = fileRecyclerView;
        this.header = header;
        this.user = user;
        this.repo = repo;
    }

    public class CodeHolder extends RecyclerView.ViewHolder {

        private LinearLayout codeViewHolder;
        private ImageView codeTypeImageView;
        private TextView codeTextView;
        private TextView codeSizeTextView;

        public CodeHolder(final View itemView) {
            super(itemView);

            codeViewHolder = itemView.findViewById(R.id.code_view_holder);
            codeTypeImageView = itemView.findViewById(R.id.code_type_image_view);
            codeTextView = itemView.findViewById(R.id.code_text_view);
            codeSizeTextView = itemView.findViewById(R.id.code_size_text_view);

            codeViewHolder.setOnClickListener(view -> {
                Code code = fileList.get(getAdapterPosition());
                if (code.getType().equals("dir")) {
                    String path = code.getPath();
                    RepoCodeFragment.repoCodeCall(fileRecyclerView, header, user, repo, path);
                    //RepoCodeFragment.displayPathAsViewObjects(path, view.getContext(), (ViewGroup) itemView.getParent());
                } else { // code.getType().equals("file)
                    Bundle fileBundle = new Bundle();
                    fileBundle.putString(BundleKeys.USER_NAME, user);
                    fileBundle.putString(BundleKeys.REPO_NAME, repo);
                    fileBundle.putString(BundleKeys.FILE_NAME, code.getName());
                    fileBundle.putString(BundleKeys.FILE_PATH, code.getPath());

                    Intent fileIntent = new Intent(itemView.getContext(), RepoFileActivity.class);
                    fileIntent.putExtras(fileBundle);
                    itemView.getContext().startActivity(fileIntent);
                }
            });
        }

    }

    @Override
    public CodeHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View codeView = inflater.inflate(R.layout.view_holder_code, parent, false);
        return new CodeHolder(codeView);
    }

    @Override
    public void onBindViewHolder(CodeHolder holder, int position) {
        Code code = fileList.get(position);

        if (code.getType().equals("dir")) {
            holder.codeTypeImageView.setImageResource(R.drawable.ic_directory);
            holder.codeSizeTextView.setVisibility(View.GONE);
        } else { // code.getType().equals("file")
            holder.codeTypeImageView.setImageResource(R.drawable.ic_file);
            holder.codeSizeTextView.setText(FileSizeParser.parseSize(code.getSize()));
        }
        holder.codeTextView.setText(code.getName());
    }

    @Override
    public int getItemCount() {
        return fileList.size();
    }

}

File: app/src/main/java/mb00/android/codehub/ui/repo/adapter/CodeAdapter.kt
package mb00.android.codehub.ui.repo.adapter

import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import mb00.android.codehub.R
import mb00.android.codehub.api.model.Code
import mb00.android.codehub.data.BundleKeys
import mb00.android.codehub.logic.utils.FileSizeParser
import mb00.android.codehub.ui.repo.view.RepoCodeFragment
import mb00.android.codehub.ui.repo.view.RepoFileActivity
import mb00.android.codehub.ui.repo.viewmodel.RepoCodeViewModel
import timber.log.Timber

/**
 * A RecyclerView adapter used to display folders and files in [RepoCodeFragment]
 */

class CodeAdapter(
        private val viewModel: RepoCodeViewModel,
        private val fileList: List<Code>,
        private val codeRecyclerView: RecyclerView,
        private val header: String,
        private val user: String,
        private val repo: String
) : RecyclerView.Adapter<CodeAdapter.CodeHolder>() {

    inner class CodeHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val codeViewHolder: LinearLayout = itemView.findViewById(R.id.code_view_holder)
        val codeTypeImageView: ImageView = itemView.findViewById(R.id.code_type_image_view)
        val codeTextView: TextView = itemView.findViewById(R.id.code_text_view)
        val codeSizeTextView: TextView = itemView.findViewById(R.id.code_size_text_view)

        init {
            codeViewHolder.setOnClickListener {
                val code = fileList[adapterPosition]
                if (code.type == "dir") {
                    val path = code.path
                    viewModel.loadRepoCode(header, user, repo, path!!)
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe({ repoCodeList ->
                                if (repoCodeList.isNotEmpty()) {
                                    val sortedCodeList = viewModel.sortCodeList(repoCodeList)
                                    codeRecyclerView.adapter = CodeAdapter(viewModel, sortedCodeList, codeRecyclerView, header, user, repo)
                                } else {
                                    val noRepoCodeTextView = (codeRecyclerView.parent as View).findViewById<TextView>(R.id.no_repo_code_text_view)
                                    noRepoCodeTextView.visibility = View.VISIBLE
                                }
                            }, { error -> Timber.e(error.message) })
                    //RepoCodeFragment.displayPathAsViewObjects(path, view.getContext(), (ViewGroup) itemView.getParent());
                } else { // code.getType().equals("file)
                    val fileBundle = Bundle()
                    fileBundle.putString(BundleKeys.USER_NAME, user)
                    fileBundle.putString(BundleKeys.REPO_NAME, repo)
                    fileBundle.putString(BundleKeys.FILE_NAME, code.name)
                    fileBundle.putString(BundleKeys.FILE_PATH, code.path)

                    val fileIntent = Intent(itemView.context, RepoFileActivity::class.java)
                    fileIntent.putExtras(fileBundle)
                    itemView.context.startActivity(fileIntent)
                }
            }
        }

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CodeHolder {
        val inflater = LayoutInflater.from(parent.context)
        val codeView = inflater.inflate(R.layout.view_holder_code, parent, false)
        return CodeHolder(codeView)
    }

    override fun onBindViewHolder(holder: CodeHolder, position: Int) {
        val code = fileList[position]

        if (code.type == "dir") {
            holder.codeTypeImageView.setImageResource(R.drawable.ic_directory)
            holder.codeSizeTextView.visibility = View.GONE
        } else { // code.getType().equals("file")
            holder.codeTypeImageView.setImageResource(R.drawable.ic_file)
            holder.codeSizeTextView.text = FileSizeParser.parseSize(code.size)
        }
        holder.codeTextView.text = code.name
    }

    override fun getItemCount(): Int {
        return fileList.size
    }

}
